Switch to one Participant per Context

ivanpauno opened the pull request:
API changes needed after https://github.com/ros2/rmw_fastrtps/pull/312.

hidmic commented on rcl/test/CMakeLists.txt:
```diff
@@ -98,6 +98,7 @@ function(test_target_function)
     APPEND_LIBRARY_DIRS ${extra_lib_dirs}
     LIBRARIES ${PROJECT_NAME}
     AMENT_DEPENDENCIES ${rmw_implementation} "osrf_testing_tools_cpp" "test_msgs"
+    TIMEOUT 120
```
@ivanpauno what happened here?

ivanpauno replied:
I saw it was flaky because of timeout, and I aummented it.

fujitatomoya commented on rcl/include/rcl/localhost.h:
```diff
@@ -22,18 +22,24 @@ extern "C"
- * variable
- * \returns true if ROS_LOCALHOST_ONLY is set and is 1, false otherwise.
+ * Checks if localhost should be used for network communication.
+ * If `localhost_only` is RCL_LOCALHOST_ONLY_DEFAULT, ROS_LOCALHOST_ONLY env variable is used.
```
```suggestion
 * If `localhost_only` is RMW_LOCALHOST_ONLY_DEFAULT, ROS_LOCALHOST_ONLY env variable is used.
```

ivanpauno commented:
@kyrofa @ruffsl @mikaelarguedas In commit https://github.com/ros2/rcl/pull/515/commits/c2710820b6c6ca3e3d99d7b11c7d52bf3c78d974, I modified how the security directory lookup work when a participant is mapped to a node. I would like to get feedback on the approach.
@ruffsl any news about the changes in `sros2`?

I haven't extended the context name API in uper layers yet, though that should be strightforward.

I left the `security_options` argument in `rmw_create_node`, just to made review simpler, but it'll be ignored except a `Participant` is mapped to a `Node`.
I think that in a future iteration, we can consider removing it.
For rmw implementations that map one participant to a node, the security path stored in the [context security options](https://github.com/ros2/rmw/blob/8254c7f9166bdc354309143d187428536989ff3a/rmw/include/rmw/init_options.h#L54) can be used as a root and then do the lookup based in node name and namespace (though that would imply moving the lookup code out of rcl).

Another thing, I'm ignoring the `ROS_SECURITY_NODE_DIRECTORY` environment variable, because it doesn't make much sense when a `Participant` is mapped to a `Context`. I'm not sure if somebody was using it before, if not I think we should consider to depecrate it.

----

There's not much changes in the [rmw_fastrtps PR] about security, I just added a function that indicates if using the node name or the context name.
In the [rmw PR](https://github.com/ros2/rmw/pull/189), I've renamed `rmw_node_security_options_t` to `rmw_security_options_t` and added a few new utility functions.

kyrofa commented on rcl/src/rcl/init.c:
```diff
@@ -144,6 +148,60 @@ rcl_init(
+  RCUTILS_LOG_DEBUG_NAMED(
+    ROS_PACKAGE_NAME, "Using security: %s", use_security ? "true" : "false");
+
+  if (!rmw_use_node_name_in_security_directory_lookup()) {
```
So if the RMW sets this to be true, this chunk of code doesn't do anything and the old behavior in rcl_node_init() takes effect, right? It makes for what looks like quite a bit of duplication between these two functions-- is there a reason they can't be closer to each other? Or at least could we refactor this process into a single function call that does the right thing depending on RMW support?

ivanpauno replied:
Yes, I think I can avoid most of the duplication.

ivanpauno replied:
Done in https://github.com/ros2/rcl/pull/515/commits/b6ed0b46706415ea063b697da30f75cb8c9749fa.

ivanpauno commented on rcl/include/rcl/security.h:
```diff
@@ -0,0 +1,104 @@
+ * (by performing longest-prefix matching).
+ *
+ * However, this expansion can be overridden by setting the secure node directory environment
+ * variable (only in rmw implementations that associate a security directory with a node),
```
I think that it also didn't make much sense for nodes.
I mean, you could pass a different `ROS_SECURITY_DIRECTORY` to each process, and that would work too (the files would need to be in `$ROS_SECURITY_DIRECTORY/node_namespace/node_name` instead of `ROS_SECURITY_NODE_DIRECTORY`).
If the intention was to specify the same security directory for many nodes in the same process, it wouldn't make much sense to do the same for context, as you can just use the same context name (node names are supposed to be unique).

In general, I try to avoid features that can cause confusion.
Using `ROS_SECURITY_NODE_DIRECTORY` when you have one `Participant` per `Context` would be confusing. A name for the environment variable like `ROS_SECURITY_CONTEXT_DIRECTORY` would make more sense, but I don't see the value of such an option.

kyrofa replied:
Yeah I agree, just makes me nervous to remove stuff I don't think makes sense, as I may not fully understand the problem the feature is trying to solve. Maybe @ruffsl or @mikaelarguedas will have some insight to share, here.

kyrofa replied:
Yeah I agree, just makes me nervous to remove stuff I don't think makes sense, as I may not fully understand the problem the feature is trying to solve. Maybe @ruffsl or @mikaelarguedas will have some insight to share, here.

mikaelarguedas replied:
The main use I have for this is when I want to apply a single set of artifacts/policies to an entire subsystem.
e.g. "let's run this entire system with encryption and no access control and see how it performs"

Without this option we'd need to set a different environment with an appropriate env variable for each and every process that your launchfile (and included launchfiles) will start (that means a lot of launchfile editing)

Another example is to have an "admin" set of artifacts that can publish/subscribe to everything. You can point all your debugging tools to this single directory and not have to remap all your debug nodes / contexts to have the same name or prefix to fit the matching criteria.

---

So overall, even if it's not meant for use by default, I'm in favor of keeping such a feature

kyrofa replied:
> You can point all your debugging tools to this single directory and not have to remap all your debug nodes / contexts to have the same name or prefix to fit the matching criteria.

I think this use-case might be better solved by having the debugging tools setup to use a common context name that can get a set of predefined, wide-open profiles (this can be disabled for production, of course). @SidFaber has done a lot of thinking about that.

ivanpauno replied:
> Without this option we'd need to set a different environment with an appropriate env variable for each and every process that your launchfile (and included launchfiles) will start (that means a lot of launchfile editing)

Maybe we could rename the variable to `ROS_SECURITY_DIRECTORY_OVERRIDE`, and avoid using `NODE` on it. We should keep backwards compatibility in rmw implementations that can actually set a different security directory for each point, and still accept the old name.

There's also another way to solve this. 99% of the processes, will have just one `Context` per process and use the default name on it. In that case, all launched processes will use the same security artifacts/policies. The only exception when you wouldn't use the default name, is when you need to create more than one (bridge between different domains, and other super stranges use cases). In those cases, you actually need to use a different context name, if not you won't be able to remap them separatly. For those exceptions, if you want to use the same security artifacts, you will have to remap both names againt to the "default context name".

ivanpauno replied:
@kyrofa @mikaelarguedas Friendly ping.

kyrofa replied:
Sorry for the delay @ivanpauno, we've been chatting about exactly this in the WG. I promise we're not ignoring you!

kyrofa replied:
The general feeling is that this variable does indeed serve a purpose. I suggest keeping it, although I'm in favor of @ivanpauno's suggested rename.

ruffsl replied:
> you could pass a different ROS_SECURITY_DIRECTORY to each process

Having to specify an entirely different keystore could be tedious in the sense that one would have to take care that the two keystores shared the correct common certificate authorities and governances. I guess that isn't too difficult do one off, would currently consist of copying over the root level keystore files; though keeping the two keystores in sync would be fragile.

> I'm not sure if somebody was using it before

Here are some links to its purpose and origin:

* https://github.com/ros2/sros2/issues/69
* https://github.com/ros2/rcl/pull/300
* a82793ce3c70a9dc80fada177d01ac9181edfea8

> If the intention was to specify the same security directory for many nodes in the same process, it wouldn't make much sense to do the same for context, as you can just use the same context name
>
> ---
> We should keep backwards compatibility in rmw implementations that can actually set a different security directory for each point, and still accept the old name.

Hmm, I think I may have been under the impression that mapping multiple nodes per process to one context would remain optional, thus still warenting the use of `ROS_SECURITY_NODE_DIRECTORY`. However, reading further into this PR, it seems that the old style of mapping has been removed entirely (which I'm ok with). This may make it tricky to measure the minimal span of permissions required for each node that are never launched as their own process, as permissions brought in by other nodes could implicitly fill in the gaps of other node policies, akin to underspecified dependencies in package.xml manifest. But perhaps this just necessitates better measurement tooling of requested graph resources in SROS.

> There's also another way to solve this. 99% of the processes, will have just one Context per process and use the default name on it.

To be clear, while the norm may be for most processes to consist of only one context, different processes should not necessarily share a `default` context directory. Unless of course the user is unconcerned about access control, or the granularity of concern is that of a robot, and not a particular process in the robot. From a security perspective, our goal is to minimize the the attack surface or privilege of each process in the graph to what is only essential for it's application.

> For those exceptions, if you want to use the same security artifacts, you will have to remap both names againt to the "default context name".

Could you clarify by what you mean here? When you have a bridging scenario of one process, consisting of two separate contexts, say for bridging over two different DDS domains? If the two DDS domains where mutually exclusive, e.g. the ranges of domains expected to apply different permissions didn't overlap, then one could just use one context (one set of DDS security artifacts) as the DDS policy schema has it's primary pivot of allowed/denied permissions on the domain range.

https://github.com/ros2/sros2/blob/fc777a2b6957cc7481341bd9f7f49e7538035e9b/sros2/test/policies/permissions.xml#L10-L12

Going back to @mikaelarguedas https://github.com/ros2/rcl/pull/515#discussion_r372643420 , I think as long as users have a method for manually overriding what context directory gets loaded via the environment, for prototyping, deguing, and auditing, then that would serve the same purpose.

ivanpauno replied:
> Having to specify an entirely different keystore could be tedious in the sense that one would have to take care that the two keystores shared the correct common certificate authorities and governances. I guess that isn't too difficult do one off, would currently consist of copying over the root level keystore files; though keeping the two keystores in sync would be fragile.

:+1: 

> However, reading further into this PR, it seems that the old style of mapping has been removed entirely (which I'm ok with).

Yes, the implementations will be refactored to map a `Participant` to a `Context`.

> Could you clarify by what you mean here? When you have a bridging scenario of one process, consisting of two separate contexts, say for bridging over two different DDS domains? If the two DDS domains where mutually exclusive, e.g. the ranges of domains expected to apply different permissions didn't overlap, then one could just use one context (one set of DDS security artifacts) as the DDS policy schema has it's primary pivot of allowed/denied permissions on the domain range.

Yes, I was refering to bridging two domains. In that case, you need two Contexts. as you need one participant for each domain. If they can share the security artifacts, both of them could use the same context name.

> Going back to @mikaelarguedas #515 (comment) , I think as long as users have a method for manually overriding what context directory gets loaded via the environment, for prototyping, deguing, and auditing, then that would serve the same purpose.

I will add `ROS_SECURITY_DIRECTORY_OVERRIDE`, which is a clearer name.

ivanpauno replied:
Solved in https://github.com/ros2/rcl/pull/515/commits/d284a2e92ac07c3afc362b802d8b2851e302fbec.

mikaelarguedas commented on rcl/src/rcl/init.c:
```diff
@@ -144,6 +148,60 @@ rcl_init(
+  RCUTILS_LOG_DEBUG_NAMED(
+    ROS_PACKAGE_NAME, "Using security: %s", use_security ? "true" : "false");
+
+  if (!rmw_use_node_name_in_security_directory_lookup()) {
```
Could this be deduplicated? looks like a lot of copy-n-paste happened here

ivanpauno replied:
Deduplicated code in https://github.com/ros2/rcl/pull/515/commits/b6ed0b46706415ea063b697da30f75cb8c9749fa.

mikaelarguedas commented on rcl/src/rcl/node.c:
```diff
@@ -258,79 +256,57 @@ rcl_node_init(
-  } else {
-    domain_id = node->impl->options.domain_id;
+  domain_id = node->impl->options.domain_id;
+  if (RMW_RET_OK != rcl_domain_id(&domain_id)) {
```
RCL_RET_OK ?

ivanpauno replied:
Thanks! Corrected in https://github.com/ros2/rcl/pull/515/commits/0317e66b33fd479dddfc38a8297a31d8801957a3.

mikaelarguedas commented on rcl/src/rcl/security.c:
```diff
@@ -0,0 +1,310 @@
+ *   which is currently a path to a directory on the filesystem containing DDS Security permission files.
+ */
+typedef char * (* security_lookup_fn_t) (
+  const char * node_name,
```
If we were to replace node everywhere, all this file could be updated as well.
Though it's unclear what word to use to capture "either context or node". I mean "participant" but it's not a ROS concept per say

ivanpauno replied:
Avoided to mention node as far as possible.

mikaelarguedas commented on rcl/src/rcl/security.c:
```diff
@@ -0,0 +1,310 @@
+ * _TINYDIR_FILENAME_MAX characters.
+ * \return true if a match was found
+ */
+static bool get_best_matching_directory(
```
Is this file a copy of `security_directory.c` ?
Is there a reason to duplicate the files?

If no need to keep the old file, could we just rename it and preserve history instead

ivanpauno replied:
Oh, I thought I renamed (I actually renamed in the `CMakeLists.txt`, and `security_directory` is not compiled more).
Don't remember what I did, I will try to clean the history.

ivanpauno replied:
I've cleaned the history now.

ivanpauno commented on rcl/include/rcl/security.h:
```diff
@@ -0,0 +1,104 @@
+ * Such an override is useful for where the FQN of a node is non-deterministic before runtime,
+ * or when testing and using additional tools that may not otherwise be easily provisioned.
+ *
+ * \param[in] name validated node name (a single token)
```
Removed, except from the part that's referring to the above discussion: https://github.com/ros2/rcl/pull/515#discussion_r372513435.
I will update when the discussion is solved.

ivanpauno commented on rcl/include/rcl/security.h:
```diff
@@ -0,0 +1,104 @@
+ *
+ * E.g. for a node/context named "c" in namespace "/a/b", the secure root path will
+ * be "a/b/c", where the delimiter "/" is native for target file system (e.g. "\\" for _WIN32).
+ * If no exact match is found for the node/context name, a best match would be used instead
```
Done.

ivanpauno commented on rcl/include/rcl/security.h:
```diff
@@ -0,0 +1,104 @@
+ */
+RCL_PUBLIC
+rcl_ret_t
+rcl_use_security(bool * use_security);
```
Updated to `rcl_security_enabled`.

ivanpauno commented on rcl/src/rcl/init.c:
```diff
@@ -141,6 +145,60 @@ rcl_init(
+    goto fail;
+  }
+
+  // TODO(ivanpauno): Uncomment this when new path discovery magic is ready.
```
Ups, forgot to delete that.
It was referring to the directory lookup, I borrowed the comment from here https://github.com/ros2/rcl/blob/7537f8304493e9b9568748b5601f5c1b7fa463f0/rcl/src/rcl/node.c#L319.

ivanpauno commented:
I think I addressed all the comments, except https://github.com/ros2/rcl/pull/515#discussion_r372513435 where we need to reach concensous.
I've also to clean history, as commented in https://github.com/ros2/rcl/pull/515#discussion_r372635815.

hidmic commented on rcl/src/rcl/security.c:
```diff
@@ -0,0 +1,353 @@
+  const rcl_allocator_t * allocator
+);
+
+security_lookup_fn_t g_security_lookup_fns[] = {
```
@ivanpauno can be made `const`?

hidmic commented on rcl/src/rcl/security.c:
```diff
@@ -0,0 +1,353 @@
+  ROS_SECURITY_LOOKUP_MATCH_PREFIX = 2,
+} ros_security_lookup_type_t;
+
+char * g_security_lookup_type_strings[] = {
```
@ivanpauno can be made `const`?

jacobperron commented on rcl/include/rcl/security.h:
```diff
@@ -0,0 +1,139 @@
+ */
+RCL_PUBLIC
+rcl_ret_t
+rcl_security_enabled(bool * use_security);
```
Why not just return a `bool`?

ivanpauno replied:
To differentiate an error condition from security being enbled or not.

jacobperron replied:
:ok_hand:

jacobperron commented on rcl/src/rcl/security.c:
```diff
@@ -0,0 +1,356 @@
+    }
+
+    if (!use_security) {
+      security_options->enforce_security = RMW_SECURITY_ENFORCEMENT_PERMISSIVE;
```
We can return early here and avoid the `else`. Also, I think we can move this check to before L54, right?

ivanpauno replied:
Checking `rmw_use_node_name_in_security_directory_lookup` in this function was an error.
I will add a test case to cover that.

ivanpauno replied:
See https://github.com/ros2/rcl/pull/515/commits/7111a4598dbd9c37a34cb2fda0fcfa804ee2201f.

jacobperron commented on rcl/src/rcl/security.c:
```diff
@@ -0,0 +1,356 @@
+  const char * ros_security_enable = NULL;
+  const char * get_env_error_str = NULL;
+
+  if (!use_security) {
```
We have a macro for this: 

`RCL_CHECK_ARGUMENT_FOR_NULL(use_security, RCL_RET_INVALID_ARGUMENT)`

jacobperron commented on rcl/src/rcl/security.c:
```diff
@@ -0,0 +1,356 @@
+
+  if (!policy) {
+    return RCL_RET_INVALID_ARGUMENT;
+  }
```
`RCL_CHECK_ARGUMENT_FOR_NULL(policy, RCL_RET_INVALID_ARGUMENT)`

jacobperron commented on rcl/src/rcl/security.c:
```diff
@@ -0,0 +1,356 @@
+/**
+ * A security lookup function takes in the node/context's name, namespace, a security root directory and an allocator;
+ *  It returns the relevant information required to load the security credentials,
+ *   which is currently a path to a directory on the filesystem containing DDS Security permission files.
```
nit: The indentation looks wrong here.

jacobperron replied:
This change is unrelated.

jacobperron commented on rcl/src/rcl/security.c:
```diff
@@ -0,0 +1,356 @@
+  prefix_match_lookup,
+};
+
+typedef enum ros_security_lookup_type_e
```
Is there a reason to use the suffix `_e` instead of `_t` like the other enums in `rcl`?

ivanpauno replied:
This comes from the previous `security_directory.h`. I can fix it if you want, but it's unrelated.

P.S.: I don't know why github just shows a deleted file and a new file in the diff.
  If you check the diff of each commit, you would be able to understand better the history (as there is one commit just renaming files). For the same reason, I will avoid squash and merge, and left a separated commit renaming the files.

jacobperron replied:
I missed that the file was renamed. We can leave it as-is since the change is unrelated.

jacobperron commented on rcl/src/rcl/security.c:
```diff
@@ -0,0 +1,356 @@
+  "MATCH_PREFIX"
+};
+
+/// Return the directory whose name most closely matches name (longest-prefix match),
```
```suggestion
/// Return the directory that most closely matches a given name (longest-prefix match),
```

jacobperron commented on rcl/test/rcl/test_node.cpp:
```diff
@@ -120,14 +120,11 @@ TEST_F(CLASSNAME(TestNodeFixture, RMW_IMPLEMENTATION), test_rcl_node_accessors)
-  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
-  {
-    EXPECT_EQ(RCL_RET_OK, rcl_context_fini(&invalid_context)) << rcl_get_error_string().str;
-  });
```
nit: This change seems unrelated to the PR

ivanpauno replied:
Without that change, `invalid_context` may be finalized before `invalid_node`.
The documentation says that doing so is UB.
Before, doing that in the wrong order didn't create a problem, now it segfaults.

jacobperron replied:
Thanks.

ivanpauno commented on rcl/include/rcl/security.h:
```diff
@@ -0,0 +1,139 @@
+#endif
+
+#ifndef ROS_SECURITY_STRATEGY_VAR_NAME
+#define ROS_SECURITY_STRATEGY_VAR_NAME "ROS_SECURITY_STRATEGY"
```
I modified the style in this file too:
```c
#ifndef ...
# define ...
#endif
```
which we use everywhere else.

ivanpauno commented on rcl/include/rcl/security.h:
```diff
@@ -0,0 +1,139 @@
+ *  \sa rcl_get_enforcement_policy
+ *  \sa rcl_get_secure_root
+ *
+ * \param name[in] name used to find the securiy root path.
```
Good catch, thanks!

ivanpauno commented on rcl/src/rcl/domain_id.c:
```diff
@@ -0,0 +1,56 @@
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+
+  if (RCL_DEFAULT_DOMAIN_ID != *domain_id) {
```
Solved.

ivanpauno commented on rcl/src/rcl/localhost.c:
```diff
@@ -12,19 +12,39 @@
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+
+  if (RMW_LOCALHOST_ONLY_DEFAULT == *localhost_only) {
```
Solved.

kyrofa commented on rcl/src/rcl/security.c:
```diff
@@ -0,0 +1,351 @@
+  const rcl_allocator_t * allocator)
+{
+  bool ros_secure_directory_override = true;
+  bool use_node_name_in_lookup = rmw_use_node_name_in_security_directory_lookup();
```
It appears that this is not used, is that true?

ivanpauno replied:
Fixed in https://github.com/ros2/rcl/pull/515/commits/533ea2bbd7a56468cedca1323de9f5f176218920.

hidmic commented on rcl/include/rcl/security.h:
```diff
@@ -0,0 +1,138 @@
+# define ROS_SECURITY_ENABLE_VAR_NAME "ROS_SECURITY_ENABLE"
+#endif
+
+/// Init the security options from the values of the environment variables and passed names.
```
@ivanpauno nit: consider 
```suggestion
/// Initializes security options from values in the environment and given names.
```

ivanpauno replied:
I will use imperative mood instead of indicative mood, but otherwise follow your suggestion.

ivanpauno replied:
Done in https://github.com/ros2/rcl/pull/515/commits/a905ae9dc01d24d4b09f0999f538b61a0e83315e.

hidmic commented on rcl/include/rcl/security.h:
```diff
@@ -0,0 +1,138 @@
+
+/// Init the security options from the values of the environment variables and passed names.
+/**
+ * Inits the passed security options based on the environment.
```
@ivanpauno same about using `Init` as a verb.

ivanpauno replied:
Done https://github.com/ros2/rcl/pull/515/commits/a905ae9dc01d24d4b09f0999f538b61a0e83315e.

hidmic commented on rcl/src/rcl/init.c:
```diff
@@ -155,7 +193,6 @@ rcl_init(
   }
 
   TRACEPOINT(rcl_init, (const void *)context);
-
```
@ivanpauno spurious change?

ivanpauno replied:
Solved in https://github.com/ros2/rcl/pull/515/commits/a905ae9dc01d24d4b09f0999f538b61a0e83315e.

hidmic commented on rcl/src/rcl/domain_id.c:
```diff
@@ -0,0 +1,52 @@
+rcl_get_default_domain_id(size_t * domain_id)
+{
+  const char * ros_domain_id;
+  const char * get_env_error_str;
```
@ivanpauno consider zero initializing these too.

ivanpauno replied:
Done, see https://github.com/ros2/rcl/pull/515/commits/a905ae9dc01d24d4b09f0999f538b61a0e83315e.

hidmic commented on rcl/src/rcl/security.c:
```diff
@@ -0,0 +1,350 @@
+  }
+
+  RCUTILS_LOG_DEBUG_NAMED(
+    ROS_PACKAGE_NAME, "Using security: %s", use_security ? "true" : "false");
```
@ivanpauno this should be moved up, `use_security` will never be `false` at this point.

ivanpauno replied:
Good catch, thanks!

ivanpauno replied:
See https://github.com/ros2/rcl/pull/515/commits/a905ae9dc01d24d4b09f0999f538b61a0e83315e.

hidmic commented on rcl/src/rcl/security.c:
```diff
@@ -0,0 +1,350 @@
+    return RCL_RET_ERROR;
+  }
+
+  *use_security = (0 == strcmp(ros_security_enable, "true"));
```
@ivanpauno meta: uh, this environment variable expects `true` as a boolean literal, while others above expect `1`. We should probably standardize that in a follow-up PR.

ivanpauno replied:
It was already working like that. See that I first just moved code https://github.com/ros2/rcl/pull/515/commits/4aa706cf34cc69d2c86920c161c4aa964fb706a7.

I agree we should standarize. We should probably have a function in `rcutils` that accepts either "1" "true" "on" (not case sensitive).

hidmic commented on rcl/src/rcl/security.c:
```diff
@@ -0,0 +1,350 @@
+        matched_name_length > max_match_length)
+      {
+        max_match_length = matched_name_length;
+        memcpy(matched_name, file.name, max_match_length);
```
@ivanpauno same as in other PRs, use `strncpy` or `rcutils_format_string` or `rcutils_strdup` instead of plain old `memcpy`

ivanpauno replied:
I didn't modify, I just moved it.
In this case, `rcutils_strdup` doesn't apply.
I will replace `memcpy` with `strncpy`, which is more appropiate here.

ivanpauno replied:
See https://github.com/ros2/rcl/pull/515/commits/a905ae9dc01d24d4b09f0999f538b61a0e83315e.

hidmic commented on rcl/src/rcl/security.c:
```diff
@@ -0,0 +1,350 @@
+    // TODO(ros2team): remove the hard-coded value of the root namespace
+    fqn = rcutils_format_string(*allocator, "%s%s%s", namespace_, "/", name);
+    // Get native path, ignore the leading forward slash
+    // TODO(ros2team): remove the hard-coded length, use the length of the root namespace instead
```
@ivanpauno meta: these TODOs seem to have been solved.

ivanpauno replied:
I still see a "/" and a `+ 1` there, so I don't think they were solved.

hidmic replied:
Oh, I see. Nevermind.

hidmic commented on rcl/src/rcl/security.c:
```diff
@@ -0,0 +1,189 @@
+  // File discovery magic here
+  char * secure_root = rcl_get_secure_root(
+    name,
+    allocator);
```
@ivanpauno nit: is there a need to wrap these arguments?

ivanpauno replied:
See https://github.com/ros2/rcl/pull/515/commits/e3d5ebdf6094274f7f74d752ecce3fbadec84150.

hidmic commented on rcl/src/rcl/security.c:
```diff
@@ -0,0 +1,189 @@
+    }
+    if (0 == strcmp("", env_buf)) {
+      return NULL;  // environment variable was empty
+    } else {
```
@ivanpauno nit: on need for an `else` clause

ivanpauno replied:
See https://github.com/ros2/rcl/pull/515/commits/e3d5ebdf6094274f7f74d752ecce3fbadec84150.

hidmic commented on rcl/src/rcl/validate_security_context_name.c:
```diff
@@ -0,0 +1,148 @@
+  if (!security_context) {
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+  return rmw_validate_namespace_with_size(
```
@ivanpauno shouldn't this be `rcl_validate_security_context_name_with_size`?

ivanpauno replied:
See https://github.com/ros2/rcl/pull/515/commits/e3d5ebdf6094274f7f74d752ecce3fbadec84150.

hidmic commented on rcl/src/rcl/validate_security_context_name.c:
```diff
@@ -0,0 +1,148 @@
+          if (ret < 0) {
+            RCL_SET_ERROR_MSG(
+              "rcl_validate_security_context_name_with_size(): "
+              "rcutils_snprintf() failed");
```
@ivanpauno it would be good to leave a trace of the original error, e.g. 
```suggestion
              "rcutils_snprintf() failed while reporting an unknown validation result");
```

ivanpauno replied:
See https://github.com/ros2/rcl/pull/515/commits/e3d5ebdf6094274f7f74d752ecce3fbadec84150.

hidmic commented on rcl/src/rcl/validate_security_context_name.c:
```diff
@@ -0,0 +1,148 @@
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+
+  int t_validation_result;
```
@ivanpauno what does the `t_` stand for? Why not: 
```suggestion
  int rmw_validation_result;
```
?

Same elsewhere.

ivanpauno replied:
temporal? to differentiate from the parameter being passed.

hidmic replied:
Consider using `tmp`, it's somewhat more common.

ivanpauno replied:
See https://github.com/ros2/rcl/pull/515/commits/44a15be05484088c78a4e7cde08a021356757160.

hidmic commented on rcl/src/rcl/validate_security_context_name.c:
```diff
@@ -0,0 +1,148 @@
+
+  // security_context might be longer that namespace length, check false positives and correct
+  if (t_validation_result == RMW_NAMESPACE_INVALID_TOO_LONG &&
+    security_context_length <= RCL_CONTEXT_NAME_MAX_LENGTH)
```
@ivanpauno mind to explain? Also, I have the impression some result is wrong, as this `if` clause is redundant as-is once you get to this point.

ivanpauno replied:
The security context name can be longer that a namespace.
I'm reusing the namespace check, and in the case the error was that the name was too long, I recheck if that's actually the case. If not, I correct the validation result to valid and return `RCL_RET_OK`.

Note: My logic is valid because the length is the last thing being checked:
https://github.com/ros2/rmw/blob/f76511333709a6f5d1e4543129ec73f2cd2b4f06/rmw/src/validate_full_topic_name.c#L117
https://github.com/ros2/rmw/blob/f76511333709a6f5d1e4543129ec73f2cd2b4f06/rmw/src/validate_namespace.c#L112

hidmic replied:
I may be crazy, but, let's say `tmp_validation_result` equals `RMW_NAMESPACE_INVALID_TOO_LONG` and `security_context_length` exceeds `RCL_SECURITY_CONTEXT_NAME_MAX_LENGTH`. The switch statement [will be skipped](https://github.com/ros2/rcl/pull/515/files#diff-6d1e34c077896ab6d3f8e1a936573cccR65), the  if statement immediately below [will be skipped](https://github.com/ros2/rcl/pull/515/files#diff-6d1e34c077896ab6d3f8e1a936573cccR112) as well, and the implementation will conclude that [the name is valid](https://github.com/ros2/rcl/pull/515/files#diff-6d1e34c077896ab6d3f8e1a936573cccR120).

ivanpauno replied:
You're right, I failed to get your point the first time.
See https://github.com/ros2/rcl/pull/515/commits/675e4d566319dc784cd514fab12bb2858f34a1f7.

hidmic commented on rcl/test/rcl/test_security.cpp:
```diff
@@ -0,0 +1,221 @@
+#ifdef _WIN32
+  return _putenv(env_var);
+#else
+  return putenv(reinterpret_cast<char *>(const_cast<char *>(env_var)));
```
Hmm, why an (apparently redundant) `reinterpret_cast` here?

ivanpauno replied:
good question, it's preexistence but I will update.

ivanpauno replied:
See https://github.com/ros2/rcl/pull/515/commits/e3d5ebdf6094274f7f74d752ecce3fbadec84150.

hidmic commented on rcl/test/rcl/test_security.cpp:
```diff
@@ -0,0 +1,221 @@
+    allocator = rcl_get_default_allocator();
+    root_path = nullptr;
+    secure_root = nullptr;
+    base_lookup_dir_fqn = nullptr;
```
@ivanpauno nit: any reason not to directly initialize these member fields?

ivanpauno replied:
Preexistent. Some tests set them to different things.

hidmic commented on rcl/test/rcl/test_security.cpp:
```diff
@@ -0,0 +1,221 @@
+    OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+    {
+      allocator.deallocate(root_path, allocator.state);
+    });
```
@ivanpauno why do you need `OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT` here? `rcl`'s default allocator will never throw.

hidmic replied:
I personally think it'd make tests clearer if these deallocations were right next to the allocation spot. I know that's duplicating, but it makes them easier to read.

ivanpauno replied:
preexistent

hidmic commented on rcl/test/rcl/test_security.cpp:
```diff
@@ -0,0 +1,221 @@
+  std::string secure_root_str(secure_root);
+  ASSERT_STREQ(
+    TEST_SECURITY_CONTEXT,
+    secure_root_str.substr(secure_root_str.size() - (sizeof(TEST_SECURITY_CONTEXT) - 1)).c_str());
```
@ivanpauno I think you meant `strlen`.

ivanpauno replied:
:sweat_smile:

ivanpauno replied:
See https://github.com/ros2/rcl/pull/515/commits/e3d5ebdf6094274f7f74d752ecce3fbadec84150.

hidmic commented on rcl/test/rcl/test_security.cpp:
```diff
@@ -0,0 +1,221 @@
+   * if the node override is invalid. */
+  putenv_wrapper(
+    ROS_SECURITY_DIRECTORY_OVERRIDE
+    "=TheresN_oWayThi_sDirectory_Exists_hence_this_would_fail");
```
@ivanpauno nit: 
```suggestion
    "=TheresN_oWayThi_sDirectory_Exists_hence_this_should_fail");
```

ivanpauno replied:
See https://github.com/ros2/rcl/pull/515/commits/e3d5ebdf6094274f7f74d752ecce3fbadec84150.

hidmic commented on rcl/test/rcl/test_security.cpp:
```diff
@@ -0,0 +1,221 @@
+};
+
+TEST_F(TestGetSecureRoot, failureScenarios) {
+  ASSERT_EQ(
```
@ivanpauno here and elsewhere, do we have to `assert` or would an `expect` perform equally well? Particularly in those tests that currently have a single assertion.

ivanpauno replied:
See https://github.com/ros2/rcl/pull/515/commits/e3d5ebdf6094274f7f74d752ecce3fbadec84150.

hidmic commented on rcl/test/rcl/test_security.cpp:
```diff
@@ -0,0 +1,221 @@
+  /* Setting root dir has no effect */
+  putenv_wrapper(ROS_SECURITY_DIRECTORY_OVERRIDE "=" TEST_RESOURCES_DIRECTORY);
+  root_path = rcl_get_secure_root(
+    "name shouldn't matter", &allocator);
```
@ivanpauno nit: do we need to wrap these arguments?

ivanpauno replied:
See https://github.com/ros2/rcl/pull/515/commits/e3d5ebdf6094274f7f74d752ecce3fbadec84150.

ivanpauno commented on rcl/src/rcl/init.c:
```diff
@@ -144,6 +153,72 @@ rcl_init(
+  }
+  if (RCL_CONTEXT_NAME_VALID != validation_result) {
+    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "rcl_validate_security_context_name result is not valid: '%s'. Invalid index: %zu",
```
See https://github.com/ros2/rcl/pull/515/commits/e3d5ebdf6094274f7f74d752ecce3fbadec84150.

hidmic commented on rcl/include/rcl/graph.h:
```diff
@@ -440,6 +441,39 @@ rcl_get_node_names(
 
+/// Return a list of available nodes in the ROS graph, including their security context names.
+/**
+ * \sa Works like rcl_get_node_names, but it also include the security context name the node is
```
@ivanpauno `\sa` expects references as arguments.

Maybe you just want something like:
```suggestion
 * \sa An \ref rcl_get_node_names equivalent, but including in its output the security context name the node is using.
```
?

ivanpauno replied:
Yeah, I was trying to do that. Thanks for the suggestion!

ivanpauno replied:
See https://github.com/ros2/rcl/pull/515/commits/44a15be05484088c78a4e7cde08a021356757160

hidmic commented on rcl/include/rcl/graph.h:
```diff
@@ -440,6 +441,39 @@ rcl_get_node_names(
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Maybe [1]
+ * <i>[1] implementation may need to protect the data structure with a lock</i>
```
@ivanpauno what implementations? This may need some qualification.

ivanpauno replied:
Preexisting, see https://github.com/ros2/rcl/blob/f242967bf024804b3ed1e8b00ebf3b3b3c9541cd/rcl/include/rcl/graph.h#L456 and documentation of the other functions too.

IMO, the comment is clear.
It means that it might lock or not depending of the rmw implementation, and that the user shouldn't make any assumption.

hidmic replied:
I'd be explicit and `RMW`, but fine with it for now.

ivanpauno replied:
AFAIK, every of them are locking now (at least, the three default implementations).
It's pretty hard to have a complete list, considering that there are also external rmw implementations.

hidmic replied:
Oh, for sure, I wouldn't mention every existing RMW implementation, I'd just qualify the noun as in:
```suggestion
 * <i>[1] RMW implementation in use may need to protect the data structure with a lock</i>
```

ivanpauno replied:
Done in https://github.com/ros2/rcl/pull/515/commits/35ec36eb170bb5801c17f03637e2a9b97f13ded7.

hidmic commented on rcl/include/rcl/security.h:
```diff
@@ -0,0 +1,123 @@
+ * However, this expansion can be overridden by setting the secure directory override environment
+ * (`ROS_SECURITY_DIRECTORY_OVERRIDE`) variable, allowing users to explicitly specify the exact secure
+ * root directory to be utilized.
+ * Such an override is useful for where the security context is non-deterministic before runtime,
```
@ivanpauno nit: 
```suggestion
 * Such an override is useful for applications where the security context is non-deterministic before runtime,
```

hidmic replied:
Also, what do you mean by `non-deterministic before runtime`? Unknown?

ivanpauno replied:
Again, this was taken from the old `security_directory.h`.
The original comment isn't clear to me.

ivanpauno replied:
Used your suggestion in https://github.com/ros2/rcl/pull/515/commits/35ec36eb170bb5801c17f03637e2a9b97f13ded7.

hidmic commented on rcl/include/rcl/validate_security_context_name.h:
```diff
@@ -0,0 +1,91 @@
+
+/// Determine if a given security context name is valid.
+/**
+ * /sa The same rules as rmw_validate_namespace are used.
```
@ivanpauno same comment as elsewhere about using `\sa` with arguments other than references. Also, that slash is backwards :sweat_smile:

ivanpauno replied:
Done in https://github.com/ros2/rcl/pull/515/commits/44a15be05484088c78a4e7cde08a021356757160.

hidmic commented on rcl/include/rcl/validate_security_context_name.h:
```diff
@@ -0,0 +1,91 @@
+/// Determine if a given security context name is valid.
+/**
+ * /sa The same rules as rmw_validate_namespace are used.
+ * The only difference is the maximum length, which can be 255 characters.
```
@ivanpauno nit: 
```suggestion
 * The only difference is in the maximum allowed length, which in this case can be up to 255 characters.
```

ivanpauno replied:
See https://github.com/ros2/rcl/pull/515/commits/44a15be05484088c78a4e7cde08a021356757160.

hidmic commented on rcl/include/rcl/validate_security_context_name.h:
```diff
@@ -0,0 +1,91 @@
+
+/// Deterimine if a given security context name is valid.
+/**
+ * This is an overload with an extra parameter for the length of security_context.
```
@ivanpauno an overload in C? Of what?

ivanpauno replied:
It's kind of metaphorical :joy:.
I copied the comment from `rmw_validate_namespace.h`, I can rephrase it if you want.

hidmic replied:
Fine with keeping the overload concept, but do add a reference to the function that is being "overloaded".

ivanpauno replied:
See https://github.com/ros2/rcl/pull/515/commits/35ec36eb170bb5801c17f03637e2a9b97f13ded7.

hidmic commented on rcl/src/rcl/arguments.c:
```diff
@@ -1098,6 +1147,7 @@ rcl_arguments_fini(
       args->impl->num_param_files_args = 0;
       args->impl->parameter_files = NULL;
     }
+    args->impl->allocator.deallocate(args->impl->security_context, args->impl->allocator.state);
```
@ivanpauno `rcl_arguments_copy` is lacking code to handle this new member field.

ivanpauno replied:
Solved in https://github.com/ros2/rcl/pull/515/commits/44a15be05484088c78a4e7cde08a021356757160.

hidmic commented on rcl/src/rcl/init.c:
```diff
@@ -144,6 +153,72 @@ rcl_init(
+  }
+  if (RCL_SECURITY_CONTEXT_NAME_VALID != validation_result) {
+    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "security context name is not valid: '%s'. Invalid index: %zu",
```
@ivanpauno nit^N: capitalize, as in 
```suggestion
      "Security context name is not valid: '%s'. Invalid index: %zu",
```

ivanpauno replied:
See https://github.com/ros2/rcl/pull/515/commits/44a15be05484088c78a4e7cde08a021356757160.

hidmic commented on rcl/src/rcl/validate_security_context_name.c:
```diff
@@ -0,0 +1,148 @@
+  {
+    *validation_result = RCL_SECURITY_CONTEXT_NAME_VALID;
+    return RCL_RET_OK;
+  }
```
@ivanpauno Hmm I though I had brought up this. What if it is too long and it exceeds `RCL_SECURITY_CONTEXT_NAME_MAX_LENGTH`?

ivanpauno replied:
See https://github.com/ros2/rcl/pull/515#discussion_r395865974.

hidmic commented on rcl/test/rcl/test_get_node_names.cpp:
```diff
@@ -162,3 +163,134 @@ TEST_F(CLASSNAME(TestGetNodeNames, RMW_IMPLEMENTATION), test_rcl_get_node_names)
+  const char * node5_name = "node1";
+  const char * node5_namespace = "/";
+  rcl_node_options_t node5_options = rcl_node_get_default_options();
+  ret = rcl_node_init(&node5, node5_name, node5_namespace, &context, &node5_options);
```
@ivanpauno hmm, is this checking that the feature works with duplicated FQN/security context tuples?

ivanpauno replied:
yes

hidmic commented on rcl/include/rcl/validate_security_context_name.h:
```diff
@@ -44,8 +44,8 @@ extern "C"
 /**
- * /sa The same rules as rmw_validate_namespace are used.
- * The only difference is the maximum length, which can be 255 characters.
+ * \sa The same rules as \ref rmw_validate_namespace are used.
```
@ivanpauno I think you want to drop `\sa` here.

ivanpauno replied:
See https://github.com/ros2/rcl/pull/515/commits/77aad14964b484c6091bebe0cec28169546303a9.

hidmic commented on rcl/src/rcl/arguments.c:
```diff
@@ -1100,6 +1100,15 @@ rcl_arguments_copy(
+    if (RCL_RET_OK != rcl_arguments_fini(args_out)) {
+      RCL_SET_ERROR_MSG("Error while finalizing arguments due to another error");
+    }
+    RCL_SET_ERROR_MSG("Error while copying security context argument");
```
@ivanpauno won't you get a complain about the error message being overwritten if `rcl_arguments_fini` fails?

ivanpauno replied:
Done in https://github.com/ros2/rcl/pull/515/commits/77aad14964b484c6091bebe0cec28169546303a9.

hidmic commented on rcl/include/rcl/validate_security_context_name.h:
```diff
@@ -64,10 +64,10 @@ rcl_validate_security_context_name(
 /**
- * This is an overload with an extra parameter for the length of security_context.
- * \param[in] security_context The number of characters in security_context.
+ * \sa This is an overload of \ref rcl_validate_security_context_name with an extra parameter
```
@ivanpauno same, I think you want to drop `\sa` here.

ivanpauno replied:
Solved in https://github.com/ros2/rcl/pull/515/commits/77aad14964b484c6091bebe0cec28169546303a9.
